scripts/portal-fetcher/openshift-docs-downloader.py

```diff
@@ -0,0 +1,1186 @@
+#!/usr/bin/env python3
+"""
+OpenShift Documentation Downloader - Functional Programming Version
+
+Downloads the HTML-single version of Red Hat OpenShift documentation for a specified version,
+preserving directory structure and maintaining a database of downloaded files.
+Supports incremental updates and advanced verification.
+
+This version is designed with functional programming principles:
+- Pure functions that avoid side effects where possible
+- Immutable data structures
+- Function composition
+- Higher-order functions
+"""
+
+import argparse
+import asyncio
+import aiohttp
+import os
+import json
+import functools
+import logging
+import time
+import datetime
+from urllib.parse import urljoin, urlparse
+from bs4 import BeautifulSoup
+from pathlib import Path
+import sqlite3
+
+# Configure logging
+logging.basicConfig(
+    level=logging.INFO,
+    format='%(asctime)s - %(levelname)s - %(message)s',
+    handlers=[
+        logging.FileHandler("openshift-downloader.log"),
+        logging.StreamHandler()
+    ]
+)
+logger = logging.getLogger(__name__)
+
+# Pure functions for URL handling
+
+def normalize_url(url):
+    """
+    Normalize a URL by removing fragment identifiers and ensuring consistency.
+    
+    Args:
+        url (str): URL to normalize
+        
+    Returns:
+        str: Normalized URL
+    """
+    # Parse the URL
+    parsed = urlparse(url)
+    
+    # Remove fragment identifier
+    parsed = parsed._replace(fragment='')
+    
+    # Build the path without trailing index if present
+    path = parsed.path
+    if path.endswith('/index'):
```

Comment:
https://docs.python.org/3/library/stdtypes.html#str.removesuffix would be clearer.